File: sobit_education_library/example/test_grasp.cpp
#include "ros/ros.h"
#include <sobit_education_library/sobit_education_controller.hpp>
#include <sobit_education_library/sobit_turtlebot_controller.hpp>

int main(int argc, char *argv[]) {
    ros::init(argc, argv, "sobit_edu_grasp");

    sobit_education::SobitEducationController edu_arm_ctr;
    sobit_education::SobitEducationController edu_wheel_ctr;

    double grasp_flag = false;
    std::string target_name = "pringle";

    // Pose: detecting_pose
    std::cout << "starting" << std:: endl;
    edu_arm_ctr.moveToPose("detecting_pose");
    ros::Duration(3.0).sleep();

    // Open the hand
    edu_arm_ctr.moveJoint( sobit_education::Joint::HAND_MOTOR_JOINT, 1.57, 2.0, true );
    ros::Duration(2.0).sleep();

    // Move the hand towards the target "target_name"
    // grasp_flag = edu_arm_ctr.moveGripperToTarget(target_name, 0.0, 0.0, 0.0)
    // Caution: We shift the x axis target posiition -0.3 to avoid collision
    grasp_flag = edu_arm_ctr.moveGripperToTargetTF(target_name, -0.3, 0.0, 0.0);
    grasp_flag = edu_arm_ctr.moveGripperToTargetCoord(0.5, 0.5, 0.5, -0.3, 0.0, 0.0);
    std::cout << "Is grasped? " << (grasp_flag? "True":"False") << std::endl;
    ros::Duration(2.0).sleep();

    // Move forwards the robot (0.3m)
    edu_wheel_ctr.controlWheelLinear(0.3);
    ros::Duration(2.0).sleep();

    // Close hand
    edu_arm_ctr.moveJoint( sobit_education::Joint::HAND_MOTOR_JOINT, 0.0, 2.0, true );
    ros::Duration(2.0).sleep();

    // Move back the robot (-0.3m)
    edu_wheel_ctr.controlWheelLinear(-0.3);
    ros::Duration(2.0).sleep();

    // Pose: "initial pose"
    edu_arm_ctr.moveToPose( "initial_pose" );

    return 0;
}

File: sobit_education_library/example/test_grasp.py
#!/usr/bin/env python3
#coding: utf-8

import rospy
from sobit_education_module import SobitEducationController
from sobit_education_module import SobitTurtlebotController
from sobit_education_module import Joint
import sys

def test(target_name):
    rospy.init_node('test')
    args = sys.argv
    edu_arm_pantilt_ctr = SobitEducationController(args[0]) # args[0] : C++上でros::init()を行うための引数
    edu_wheel_ctr = SobitTurtlebotController(args[0])       # args[0] : C++上でros::init()を行うための引数

    grasp_flag = False

    # ”detecting_pose”など決められたポーズをする
    edu_arm_pantilt_ctr.moveToPose( "detecting_pose" )
    rospy.sleep(3.0)

    # ハンドを開く
    edu_arm_pantilt_ctr.moveJoint( Joint.HAND_MOTOR_JOINT, 1.57, 2.0, True )
    rospy.sleep(2.0)

    # 把持する対象の物体があった場合、そこの位置までアームを移動させる
    # 注意：x_shift=-0.3しておくと、衝突を避けられる対策である
    # grasp_flag = edu_arm_pantilt_ctr.moveGripperToTargetTF(target_name, -0.3, 0.0, 0.0)
    grasp_flag = edu_arm_pantilt_ctr.moveGripperToTargetCoord(0.2, 0.2, 0.75, -0.3, 0.0, 0.0)
    print("Was it grasped? ", grasp_flag)
    rospy.sleep(2.0)

    # ロボット全体を0.3m直進させる
    edu_wheel_ctr.controlWheelLinear(0.3)
    rospy.sleep(2.0)
    
    # ハンドを動かす
    edu_arm_pantilt_ctr.moveJoint( Joint.HAND_MOTOR_JOINT, 0.0, 2.0, True )
    rospy.sleep(2.0)

    # ロボット全体を0.3m戻ってくる
    edu_wheel_ctr.controlWheelLinear(-0.3)
    rospy.sleep(2.0)

    # "initail_pose"など決められたポーズをする
    edu_arm_pantilt_ctr.moveToPose( "initial_pose" )

if __name__ == '__main__':
    try:
        test("pringles")
    except rospy.ROSInterruptException: pass
